File: Design/1381. Design a Stack With Increment Operation.cpp
class CustomStack {
public:
    vector<int> stack, inc;
    int n;
    CustomStack(int maxSize) {
        n = maxSize;
    }

    void push(int x) {
        if (stack.size() == n) return;
        stack.push_back(x);
        inc.push_back(0);
    }

    int pop() {
        int i = stack.size() - 1;
        if (i < 0) return -1;
        if (i > 0) inc[i - 1] += inc[i];
        int res = stack[i] + inc[i];
        stack.pop_back();
        inc.pop_back();
        return res;
    }

    void increment(int k, int val) {
        int i = min(k, (int)stack.size()) - 1;
        if (i >= 0) inc[i] += val;
    }
};


//Deque 可以loop
class CustomStack {
public:
    // Space Complexity: O(n)
    deque<int> dq;
    int n;
    
    CustomStack(int maxSize) {
        // Time Complexity: O(1)
        n = maxSize;
    }
    
    void push(int x) {
        // Time Complexity: O(1)
        if (dq.size() == n) return;
        dq.push_back(x);
    }
    
    int pop() {
        // Time Complexity: O(1)
        if (dq.empty()) return -1;
        int x = dq.back();
        dq.pop_back();
        return x;
    }
    
    void increment(int k, int val) {
        // Time Complexity: O(k)
        for (auto it = dq.begin(); it != (k > dq.size() ? dq.end() : dq.begin() + k); ++it)
            *it += val;
    }
};

File: Design/1381. Design a Stack With Increment Operation.py
class CustomStack:

    def __init__(self, maxSize: int):
        self.s = []
        self.n = maxSize

    def push(self, x: int) -> None:
        if len(self.s) >= self.n: 
            return 
        self.s.append (x)

    def pop(self) -> int:
        if len(self.s): return self.s.pop()
        return -1

    def increment(self, k: int, val: int) -> None:
        n = len(self.s)
        for i in range(min(k, n)):
            self.s[i] += val



class CustomStack(object):

    def __init__(self, maxSize):
        """
        :type maxSize: int
        """
        self.__max_size = maxSize
        self.__stk = []

    def push(self, x):
        """
        :type x: int
        :rtype: None
        """
        if len(self.__stk) == self.__max_size:
            return
        self.__stk.append([x, 0])

    def pop(self):
        """
        :rtype: int
        """
        if not self.__stk:
            return -1
        x, inc = self.__stk.pop()
        if self.__stk:
            self.__stk[-1][1] += inc
        return x + inc

    def increment(self, k, val):
        """
        :type k: int
        :type val: int
        :rtype: None
        """
        i = min(len(self.__stk), k)-1
        if i >= 0:
            self.__stk[i][1] += val